### ISSUE TITLE
report_subject is treated as mandatory while in XSD is optional

### ISSUE THREAD

**Issue by michal-kowalczyk** [OpenLEADR/openleadr-python#37]
XSD says that reportSubject may occur 0 times in oadrReportDescription
```
	<xs:complexType name="oadrReportDescriptionType">
		<xs:annotation>
			<xs:documentation>Describes the subject and attributes of a report.</xs:documentation>
		</xs:annotation>
		<xs:sequence>
			<xs:element ref="ei:rID"/>
			<xs:element ref="ei:reportSubject" minOccurs="0"/>
			<xs:element ref="ei:reportDataSource" minOccurs="0"/>
			<xs:element ref="ei:reportType"/>
			<xs:element ref="emix:itemBase" minOccurs="0">
				<xs:annotation>
					<xs:documentation>What is measured or tracked in this report (Units). </xs:documentation>
				</xs:annotation>
			</xs:element>
			<xs:element ref="ei:readingType"/>
			<xs:element ref="emix:marketContext" minOccurs="0"/>
			<xs:element ref="oadr:oadrSamplingRate" minOccurs="0"/>
		</xs:sequence>
	</xs:complexType>
```

But the openleadr uses it unconditionally:
https://github.com/OpenLEADR/openleadr-python/blob/main/openleadr/service/report_service.py#L87
https://github.com/OpenLEADR/openleadr-python/blob/main/openleadr/service/report_service.py#L96

When that item is missing in the message from VTE, it results with:
```
An exception occurred during the execution of your handler: KeyError: 'report_subject'
```

**Comment by michal-kowalczyk** [comment]
Yes, indeed. Please check out the message content:
```
<?xml version="1.0" encoding="UTF-8" standalone="no" ?>
<p1:oadrPayload xmlns:p1="http://openadr.org/oadr-2.0b/2012/07">
  <p1:oadrSignedObject>
    <p1:oadrRegisterReport xmlns:p3="http://docs.oasis-open.org/ns/energyinterop/201110" p3:schemaVersion="2.0b" xmlns:p2="http://docs.oasis-open.org/ns/energyinterop/201110/payloads">
      <p2:requestID>085B862ACB</p2:requestID>
      <p1:oadrReport xmlns:p3="urn:ietf:params:xml:ns:icalendar-2.0" xmlns:p4="http://docs.oasis-open.org/ns/energyinterop/201110">
        <p3:duration>
          <p3:duration>PT120M</p3:duration>
        </p3:duration>
        <p1:oadrReportDescription xmlns:p4="http://docs.oasis-open.org/ns/energyinterop/201110" xmlns:p5="http://docs.oasis-open.org/ns/emix/2011/06/power" xmlns:p6="http://docs.oasis-open.org/ns/emix/2011/06">
          <p4:rID>rid_energy_4184bb93</p4:rID>
          <p4:reportDataSource>
            <p4:resourceID>DEVICE1</p4:resourceID>
          </p4:reportDataSource>
          <p4:reportType>reading</p4:reportType>
          <p5:energyReal xmlns:p6="http://docs.oasis-open.org/ns/emix/2011/06/siscale">
            <p5:itemDescription/>
            <p5:itemUnits>Wh</p5:itemUnits>
            <p6:siScaleCode>none</p6:siScaleCode>
          </p5:energyReal>
          <p4:readingType>Direct Read</p4:readingType>
          <p6:marketContext/>
          <p1:oadrSamplingRate>
            <p1:oadrMinPeriod>PT1M</p1:oadrMinPeriod>
            <p1:oadrMaxPeriod>PT1M</p1:oadrMaxPeriod>
            <p1:oadrOnChange>false</p1:oadrOnChange>
          </p1:oadrSamplingRate>
        </p1:oadrReportDescription>
        <p1:oadrReportDescription xmlns:p4="http://docs.oasis-open.org/ns/energyinterop/201110" xmlns:p5="http://docs.oasis-open.org/ns/emix/2011/06/power" xmlns:p6="http://docs.oasis-open.org/ns/emix/2011/06">
          <p4:rID>rid_power_4184bb93</p4:rID>
          <p4:reportDataSource>
            <p4:resourceID>DEVICE1</p4:resourceID>
          </p4:reportDataSource>
          <p4:reportType>reading</p4:reportType>
          <p5:powerReal xmlns:p6="http://docs.oasis-open.org/ns/emix/2011/06/siscale">
            <p5:itemDescription/>
            <p5:itemUnits>W</p5:itemUnits>
            <p6:siScaleCode>none</p6:siScaleCode>
            <p5:powerAttributes>
              <p5:hertz>60</p5:hertz>
              <p5:voltage>120</p5:voltage>
              <p5:ac>true</p5:ac>
            </p5:powerAttributes>
          </p5:powerReal>
          <p4:readingType>Direct Read</p4:readingType>
          <p6:marketContext/>
          <p1:oadrSamplingRate>
            <p1:oadrMinPeriod>PT1M</p1:oadrMinPeriod>
            <p1:oadrMaxPeriod>PT1M</p1:oadrMaxPeriod>
            <p1:oadrOnChange>false</p1:oadrOnChange>
          </p1:oadrSamplingRate>
        </p1:oadrReportDescription>
        <p4:reportRequestID>0</p4:reportRequestID>
        <p4:reportSpecifierID>DEMO_TELEMETRY_USAGE</p4:reportSpecifierID>
        <p4:reportName>METADATA_TELEMETRY_USAGE</p4:reportName>
        <p4:createdDateTime>2020-12-15T12:55:02.000000Z</p4:createdDateTime>
      </p1:oadrReport>
      <p3:venID>ven_id</p3:venID>
    </p1:oadrRegisterReport>
  </p1:oadrSignedObject>
</p1:oadrPayload>
```
FYI: I test your VTN against that VTE: https://github.com/epri-dev/OpenADR-VEN-Library .

**Comment by stan-janssen** [comment]
Thanks for the information. I noticed that I was accessing the wrong property; the reportSubject can only ever be a endDeviceAsset, never a resourceID (according to the OpenADR specification).

I have converted the function to (carefully) use the reportDataSource element instead of the reportSubject, and I will work on expanding support for different reportDataSource elements than just resourceID, but that requires a small API change to the callback functions.
